File: algorithms/2_medium/M/02130_maximum_twin_sum_of_a_linked_list/vector.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        assert(head != nullptr);
        auto nums = vector<int>();
        struct ListNode dummy = ListNode(0);
        dummy.next = head;
        struct ListNode* slow = &dummy;
        struct ListNode* fast = &dummy;

        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
            nums.push_back(slow->val);
        }
        
        slow = slow->next;
        int max = INT_MIN;

        while (slow != nullptr) {
            const int curr_sum = nums.back() + slow->val;
            max = std::max(curr_sum, max);
            nums.pop_back();
            slow = slow->next;
        }
        
        return max;
    }
};

File: algorithms/2_medium/M/02130_maximum_twin_sum_of_a_linked_list/vector.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn pair_sum(mut head: Option<Box<ListNode>>) -> i32 {
        let mut nums: Vec<i32> = Vec::new();

        while let Some(node) = head {
            nums.push(node.val);
            head = node.next;
        }

        let mut left = 0;
        let mut right = nums.len() - 1;
        let mut max = i32::MIN;

        while left < right {
            max = max.max(nums[left] + nums[right]);
            left += 1;
            right -= 1;
        }
        
        max
    }
}
